integr8ly/grafana-operator
# INTLY-8778-Extend folder support functionality

# Description
- extend folder support functionality 
- add ability to specify custom folder name
- add deletion method for obsolete resources 

## JIRA
https://issues.redhat.com/browse/INTLY-8778

## Verification
- Run `make cluster/prepare/local`
- Add the Makefile recipe from below, or create these dashboards manually
```
.PHONY: dashboards
dashboards:
	oc apply -f deploy/examples/dashboards/DashboardWithCustomFolder.yaml -n grafana
	oc apply -f deploy/examples/dashboards/SimpleDashboard.yaml -n grafana
	oc apply -f deploy/examples/dashboards/KeycloakDashboard.yaml -n grafana

```
- Run the operator locally with `operator-sdk run local --watch-namespace=grafana` <- ensure your operator-sdk is at version 0.18.2
- Open the Grafana UI and navigate to dashboards, ensure that the `grafana` folder has two dashboards, `simple-dashboard` and `keycloak-metrics`, and the `customFolderName`, has one `dashboard dashboard-with-custom-folder`
- run `oc delete grafanadashboard dashboard-with-custom-folder -n grafana `
- ensure that the dashboard was deleted but the folder has remained
- run ` oc delete grafanadashboard simple-dashboard -n grafana` and ensure that the dashboard was deleted and that the `keycloak-dashboard` has remained in the `grafana` folder
- run `oc delete grafanadashboard keycloak-dashboard -n grafana` and ensure that both the dashboard and folder are deleted in the Grafana UI.

Any user specified folder through the spec, `customFolderName` field should remain unchanged even if it is empty, any other "managed" folder, such as the ones that are created from the namespaces of the dashboard in which the dashboard is located should be deleted when empty
